### PR TITLE
[MXNET-424] dtype option for multinomial

### PR DESCRIPTION
**Pull request by asitstands**
## Description ##

This PR adds `dtype` option to set the data type of the sample output array of `random.multinomial`, which is fixed as 'int32' in the current implementation. The default value is `int32`.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

File: src/operator/random/sample_multinomial_op.h [hunk 1]
```diff
@@ -155,9 +158,11 @@ void SampleMultinomialForward(const nnvm::NodeAttrs& attrs,
     Tensor<xpu, 1, float> uniform =
       ctx.requested[1].get_space_typed<xpu, 1, float>(Shape1(N*M), s);
     prnd->SampleUniform(&uniform, 0, 1);
-    Kernel<SampleMultinomialKernel, xpu>::Launch(
-      s, N, K, M, inputs[0].dptr<DType>(), uniform.dptr_, outputs[0].dptr<int>(),
-      param.get_prob ? outputs[1].dptr<DType>() : nullptr);
+    MSHADOW_TYPE_SWITCH(outputs[0].type_flag_, IType, {
```
Comment (piiswrong) [inline src/operator/random/sample_multinomial_op.h]: This kind of 2 layer switches is very slow to compile. Why do we need type support for output?

Comment (asitstands) [reply src/operator/random/sample_multinomial_op.h]: Sometimes the multinomial samples need further processing in floating point arithmetic, so the samples need to be copied into a new array of floating point type which slow down the training. For example, in RBM, the samples need to be applied by `linalg.gemm` which supports only floating point arrays.

Comment (piiswrong) [reply src/operator/random/sample_multinomial_op.h]: A simple cast shouldn't cost that much?

This kind of nested switches are really slow to compile and makes the binary much bigger.
We need to make sure it really justifies the cost

Comment (asitstands) [reply src/operator/random/sample_multinomial_op.h]: The binary size increases about 0.1% for both shared and static library (CUDA, CUDNN, MKL). Compiling mxnet already takes quite long time, so the relative increase of the compile time is also tiny.

I'm working with some variants of RBM and the use of `.astype('float32')` in several places increases training time over 20%. In the case of usual basic RBM, it increases about 10% of training time in my [test](https://github.com/asitstands/incubator-mxnet/blob/e96c79d84e00e5b7d26b18c8faba0daaac812f5e/example/restricted-boltzmann-machine/binary_rbm.py#L86) for mnist. Of course, it depends on the hyperparameters and data. However, I think that, in general, the cost cannot be ignored for applications using heavy Monte Carlo samplings of discrete states.

**Comment by asmushetzel** [comment]
This change would make things a bit more consistent with other samplers. For all the rest (uniform , gamma etc), we consistently use a floating point type as return value (by default 32bit) though some of them (Poisson, negative binomial) are distributions on integer values. And for the cases that I have seen where these samplers get used in practice, in fact the users needed floating-point data for further processing. 
We hardly can't change the default type of the multinomial anymore, but I think we should add floating point as a result type. 
The nested compile switches are used a lot in MXNet, not sure whether this is an issue when used also here.

File: src/operator/random/sample_multinomial_op.h [hunk 2]
```diff
@@ -67,6 +70,10 @@ inline bool SampleMultinomialOpShape(const nnvm::NodeAttrs& attrs,
   const TShape& ishape = (*in_attrs)[0];
   if (!ishape.ndim()) return false;
 
+  MSHADOW_TYPE_SWITCH(param.dtype, DType, {
+    CHECK_LE(ishape[ishape.ndim() - 1], mxnet::common::MaxIntegerValue<DType>());
```
Comment (piiswrong) [inline src/operator/random/sample_multinomial_op.h]: Need to output a message saying why it failed.

**Comment by piiswrong** [comment]
otherwise LGTM

File: src/operator/random/sample_multinomial_op.h (hunk 2 above)
Comment (asitstands) [reply src/operator/random/sample_multinomial_op.h]: I added an error message.
